File: solutions/Easy/Maximum_Depth_of_Binary_Tree_0104.py
from typing import Optional

# Definition for a binary tree node.


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        """
        Function: maxDepth
        Description: Computes the maximum depth of a binary tree using recursion.

        Parameters:
        - root (Optional[TreeNode]): The root of the binary tree.

        Returns:
        - int: The maximum depth of the binary tree.
        """
        if not root:
            return 0

        return 1 + max(self.maxDepth(root.left), self.maxDepth(root.right))


# Helper functions for testing
def list_to_binary_tree(lst):
    """Converts a level-order list to a binary tree."""
    if not lst or lst[0] is None:
        return None

    nodes = [TreeNode(val) if val is not None else None for val in lst]

    for i, node in enumerate(nodes):
        if node:
            left_idx = 2 * i + 1
            right_idx = 2 * i + 2
            if left_idx < len(lst):
                node.left = nodes[left_idx]
            if right_idx < len(lst):
                node.right = nodes[right_idx]

    return nodes[0]  # Root node


# Example Test Cases
if __name__ == "__main__":
    solution = Solution()

    # Test case 1: Standard tree
    root = list_to_binary_tree([3, 9, 20, None, None, 15, 7])
    print(solution.maxDepth(root))
    # Expected output: 3

    # Test case 2: Single node tree
    root = list_to_binary_tree([1])
    print(solution.maxDepth(root))
    # Expected output: 1

    # Test case 3: Unbalanced tree
    root = list_to_binary_tree([1, None, 2])
    print(solution.maxDepth(root))
    # Expected output: 2

    # Test case 4: Deeper unbalanced tree
    root = list_to_binary_tree([1, 2, None, 3, None, None, None, 4])
    print(solution.maxDepth(root))
    # Expected output: 4

    # Test case 5: Empty tree
    root = list_to_binary_tree([])
    print(solution.maxDepth(root))
    # Expected output: 0

    # Test case 6: Full tree
    root = list_to_binary_tree([1, 2, 3, 4, 5, 6, 7])
    print(solution.maxDepth(root))
    # Expected output: 3

    # Test case 7: Full tree with different depths
    root = list_to_binary_tree([1, 2, 2, 3, 3, 3, 3])
    print(solution.maxDepth(root))
    # Expected output: 3

    # Test case 8: Full tree with different depths
    root = list_to_binary_tree([1, 2, 2, 3, 3, 3, 3, 4])
    print(solution.maxDepth(root))
    # Expected output: 4

    print("All test cases passed!")

# Complexity Analysis
# Time Complexity: O(N), where N is the number of nodes in the binary tree. We visit each node once.
# Space Complexity: O(H), where H is the height of the binary tree. This represents the maximum depth of the recursion stack.

File: solutions/Easy/Maximum_Depth_of_Binary_Tree_0104.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val);
 *     this.left = (left===undefined ? null : left);
 *     this.right = (right===undefined ? null : right);
 * }
 */

/**
 * Function: maxDepth (Recursive DFS)
 * Description: Given the root of a binary tree, this function returns its maximum depth.
 * The maximum depth is the longest path from the root to a leaf node.
 *
 * @param {TreeNode} root - The root node of the binary tree.
 * @returns {number} - The maximum depth of the tree.
 */
const maxDepth = function(root) {
    if (!root) return 0; // Base case: If tree is empty, depth is 0
    
    return 1 + Math.max(maxDepth(root.left), maxDepth(root.right));
};

/**
 * Function: maxDepthIterative (BFS Approach)
 * Description: This function finds the maximum depth using level-order traversal (BFS).
 *
 * @param {TreeNode} root - The root node of the binary tree.
 * @returns {number} - The maximum depth of the tree.
 */
const maxDepthIterative = function(root) {
    if (!root) return 0; // Base case: Empty tree has depth 0

    let queue = [root];
    let depth = 0;

    while (queue.length > 0) {
        let levelSize = queue.length;
        depth++;

        for (let i = 0; i < levelSize; i++) {
            let currentNode = queue.shift();

            if (currentNode.left) queue.push(currentNode.left);
            if (currentNode.right) queue.push(currentNode.right);
        }
    }

    return depth;
};

// Example Test Cases
if (require.main === module) {
    const TreeNode = function(val, left = null, right = null) {
        this.val = val;
        this.left = left;
        this.right = right;
    };

    console.log("Test 1: Balanced binary tree");
    let root1 = new TreeNode(3, 
        new TreeNode(9), 
        new TreeNode(20, new TreeNode(15), new TreeNode(7))
    );
    console.log(maxDepth(root1)); // Expected output: 3
    console.log(maxDepthIterative(root1)); // Expected output: 3

    console.log("Test 2: Right-skewed tree");
    let root2 = new TreeNode(1, null, new TreeNode(2));
    console.log(maxDepth(root2)); // Expected output: 2
    console.log(maxDepthIterative(root2)); // Expected output: 2

    console.log("Test 3: Empty tree");
    console.log(maxDepth(null)); // Expected output: 0
    console.log(maxDepthIterative(null)); // Expected output: 0

    console.log("Test 4: Single node tree");
    let root3 = new TreeNode(5);
    console.log(maxDepth(root3)); // Expected output: 1
    console.log(maxDepthIterative(root3)); // Expected output: 1

    console.log("Test 5: Left-skewed tree");
    let root4 = new TreeNode(1, new TreeNode(2, new TreeNode(3, new TreeNode(4, new TreeNode(5)))));
    console.log(maxDepth(root4)); // Expected output: 5
    console.log(maxDepthIterative(root4)); // Expected output: 5

    console.log("Test 6: Larger tree with mixed structure");
    let root5 = new TreeNode(10, 
        new TreeNode(5, new TreeNode(2), new TreeNode(7)), 
        new TreeNode(15, new TreeNode(12), new TreeNode(20))
    );
    console.log(maxDepth(root5)); // Expected output: 3
    console.log(maxDepthIterative(root5)); // Expected output: 3
}

/*
Time Complexity:
- Recursive DFS: O(n) - Each node is visited once.
- Iterative BFS: O(n) - Each node is processed level by level.

Space Complexity:
- Recursive DFS: O(h) - Recursion depth (O(log n) for balanced, O(n) for skewed trees).
- Iterative BFS: O(n) - Queue stores nodes level-wise.
*/

export { maxDepth, maxDepthIterative };
